base/inference.jl

```diff
@@ -2940,6 +2948,8 @@ function ⊑(@nospecialize(a), @nospecialize(b))
     if isa(a, Conditional)
         if isa(b, Conditional)
             return issubconditional(a, b)
+        elseif isa(b, Const) && isa(b.val, Bool)
+            return maybe_extract_const_bool(a) === b.val
```

Comment:
Should add corresponding code for `Const ⊑ Conditional`.

It also seems we don't have `Bool ⊑ Conditional`, but shouldn't that be true (unless one branch is `Bottom`)?

Reply:
Frankly, I'm not sure. `Conditional` is a bit of an odd element here. Are you suggesting that:
```
for (a, b) in ((Conditional(_, T, S), Bool), (Conditional(_, T, Union{}), Const(true)))
@test a ⊑ b
@test b ⊑ a
end
```
should hold?

Reply:
After drawing out the lattice, I see you might be right. It seems `Conditional(_, T, Union{})` is "slightly smaller" than `Const(true)`.

It's unintuitive that `Const(true) ⊑ Conditional(_, T, T)` doesn't hold, which is probably what's confusing me.

Reply:
> It's unintuitive that Const(true) ⊑ Conditional(_, T, T) doesn't hold, which is probably what's confusing me.

You're right, if both branches are the same I would expect it to be equivalent to Bool in essentially every way, because that basically says, independent of the condition, that `var` is constrained by `T`, so it really should just have been done that way in the first place.

Reply:
Yes, currently, we have that `Conditional(_, T, T) === Bool` and `Conditional(_, T, S) ⊑ Bool`

This PR is proposing making `Conditional(_, T, Union{}) ⊑ Const(true)` as a special case (currently the common lattice element of those two is `Bool`, but the former also should currently never appear)

Reply:
By `===` in the first one I assume you mean `===(a,b) = a ⊑ b && b ⊑ a`. If so, is that actually true for the current code?

Reply:
It's only true in theory, since we don't check for it specifically in `⊑`. In practice, I just usually tried to avoid producing that, so that I wouldn't have to worry about it, and it doesn't have any effect to carry around this Conditional.